File: app/internal/publisher/template/java/CryptoUtils.java
package com.websmith.sdk;

import com.google.gson.JsonObject;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.time.Instant;
import java.util.Base64;
import java.util.UUID;
import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

public class CryptoUtils {

    public static String generateTimestamp() {
        return Instant.now().toString();
    }

    public static String generateNonce() {
        return UUID.randomUUID().toString();
    }

    public static String signRequest(
            JsonObject payload,
            String secret,
            String timestamp,
            String nonce,
            String method,
            String path,
            String query
    ) throws Exception {
        String payloadJson = payload.toString();
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        byte[] bodyHash = md.digest(payloadJson.getBytes(StandardCharsets.UTF_8));
        String bodyHashHex = bytesToHex(bodyHash);
        String message = method + "\n" + path + "\n" + query + "\n" + bodyHashHex + "\n" + timestamp + "\n" + nonce;
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec keySpec = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(keySpec);
        byte[] signature = mac.doFinal(message.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(signature);
    }

    public static String signMessage(String message, String secret) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        SecretKeySpec keySpec = new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8), "HmacSHA256");
        mac.init(keySpec);
        byte[] signature = mac.doFinal(message.getBytes(StandardCharsets.UTF_8));
        return Base64.getEncoder().encodeToString(signature);
    }

    public static boolean verifySignature(
            String signature,
            JsonObject payload,
            String secret,
            String timestamp,
            String nonce,
            String method,
            String path,
            String query
    ) throws Exception {
        String expected = signRequest(payload, secret, timestamp, nonce, method, path, query);
        return MessageDigest.isEqual(
            expected.getBytes(StandardCharsets.UTF_8),
            signature.getBytes(StandardCharsets.UTF_8)
        );
    }

    public static String sha256Hex(String data) throws Exception {
        MessageDigest md = MessageDigest.getInstance("SHA-256");
        byte[] hash = md.digest(data.getBytes(StandardCharsets.UTF_8));
        return bytesToHex(hash);
    }

    public static String bytesToHex(byte[] bytes) {
        StringBuilder sb = new StringBuilder();
        for (byte b : bytes) {
            sb.append(String.format("%02x", b));
        }
        return sb.toString();
    }
}

File: app/internal/publisher/template/dotnet/CryptoUtils.cs
using System;
using System.Security.Cryptography;
using System.Text;

namespace WebsmithSDK;

public static class CryptoUtils
{
    public static string GenerateTimestamp()
    {
        return DateTime.UtcNow.ToString("yyyy-MM-ddTHH:mm:ssZ");
    }

    public static string GenerateNonce()
    {
        var bytes = new byte[16];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(bytes);
        return Convert.ToHexStringLower(bytes);
    }

    public static string SignRequest(
        string payloadJson,
        string secret,
        string timestamp,
        string nonce,
        string method = "POST",
        string path = "",
        string query = "")
    {
        var bodyHash = ComputeSha256(payloadJson);
        var message = $"{method}\n{path}\n{query}\n{bodyHash}\n{timestamp}\n{nonce}";
        return ComputeHmacSha256(message, secret);
    }

    public static string SignMessage(string message, string secret)
    {
        return ComputeHmacSha256(message, secret);
    }

    public static bool VerifySignature(
        string signature,
        string payloadJson,
        string secret,
        string timestamp,
        string nonce,
        string method = "POST",
        string path = "",
        string query = "")
    {
        var expected = SignRequest(payloadJson, secret, timestamp, nonce, method, path, query);
        return CryptographicOperations.FixedTimeEquals(
            Encoding.UTF8.GetBytes(expected),
            Encoding.UTF8.GetBytes(signature));
    }

    public static string ComputeSha256(string value)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(value));
        return Convert.ToHexStringLower(bytes);
    }

    public static string ComputeHmacSha256(string message, string secret)
    {
        var keyBytes = Encoding.UTF8.GetBytes(secret);
        var messageBytes = Encoding.UTF8.GetBytes(message);
        var hash = HMACSHA256.HashData(keyBytes, messageBytes);
        return Convert.ToHexStringLower(hash);
    }

    public static string RandomString(int length = 32)
    {
        var bytes = new byte[length];
        using var rng = RandomNumberGenerator.Create();
        rng.GetBytes(bytes);
        return Convert.ToHexStringLower(bytes);
    }

    public static string EncryptData(string data, string key)
    {
        using var aes = Aes.Create();
        aes.Key = SHA256.HashData(Encoding.UTF8.GetBytes(key));
        aes.GenerateIV();
        using var encryptor = aes.CreateEncryptor();
        var plaintext = Encoding.UTF8.GetBytes(data);
        var ciphertext = encryptor.TransformFinalBlock(plaintext, 0, plaintext.Length);
        var result = new byte[aes.IV.Length + ciphertext.Length];
        Buffer.BlockCopy(aes.IV, 0, result, 0, aes.IV.Length);
        Buffer.BlockCopy(ciphertext, 0, result, aes.IV.Length, ciphertext.Length);
        return Convert.ToBase64String(result);
    }

    public static string DecryptData(string encoded, string key)
    {
        var raw = Convert.FromBase64String(encoded);
        using var aes = Aes.Create();
        aes.Key = SHA256.HashData(Encoding.UTF8.GetBytes(key));
        var iv = new byte[aes.BlockSize / 8];
        Buffer.BlockCopy(raw, 0, iv, 0, iv.Length);
        aes.IV = iv;
        using var decryptor = aes.CreateDecryptor();
        var ciphertext = new byte[raw.Length - iv.Length];
        Buffer.BlockCopy(raw, iv.Length, ciphertext, 0, ciphertext.Length);
        var plaintext = decryptor.TransformFinalBlock(ciphertext, 0, ciphertext.Length);
        return Encoding.UTF8.GetString(plaintext);
    }
}
